### DISCUSSION TITLE
Hotkey Addons

### DISCUSSION THREAD

**Discussion by TheColaber** [ScratchAddons/ScratchAddons#890] [Discussion]
It's about time we start talking about what we will do with our new settings type, "key". This was going to be used to create new addons that use hotkeys, and even add new settings for old addons that could potentially have them. I was talking to @WorldLanguages on discord and he came up with the idea of there being a new API for this idea. They also stated this in #783.

There would be a new API, something like `addon.self.addEventListener(event, callback)`, which we would use for, not only "key", but also:
- disabled (#783)
- and more (but I can't think of any)

But back to using this function for "key". The callback would probably have one parameter, info. Info would we an object that has the following items:
- key - the keys pressed
- addonID - the id of the addon that had a setting for this key
- addonSettingID - the id of the setting that uses that hotkey
- and more (but I can't think of any)

Now, we'd also need to think about what addons could use some hotkeys that could bring them to life! I see these:
- 60fps
- editor-devtools
- forum-search
- pause
- and more (but I can't think of any)

Finally, new addons, that we could use hotkeys:
- Editor Hotkeys
- Navigation Hotkeys
- and more (but I can't think of any)

So thank you for listening to my speech. I worked very hard on it. If you see anything that can be changed or added, lmk.

**Comment by pufferfish101007** [comment]
IMO, the way it should work is there should be an `addon.keys` api, with:
- `addon.keys.getKeys(key_id)` (I'm assuming they act like like settings in the addon.json?) - get the hotkey for a hotkey (that doesn't make any sense lol)
- `addon.keys.onHotKeyChange(cb(key_id,))` - like addons.settings.onChange 
- `addon.keys.onHotKey(cb(key_id))`  - when a hotkey relevant to that addon is pressed is pressed; means we don't need to worry about what keys were pressed, just the event it should trigger.

With this structure the addon doesn't need to worry about what keys are for what hotkey, and it can just stroll along with nice easy events, and it'll make all the code smaller. The only reason that we should keep the methods to get the keys is so the addons can add hints or stuff to remind people of theur hotkeys.

**Reply by TheColaber** [reply]
- `addon.keys.getKeys(key_id)` is the same as addon.settings.get(id) - so there's no point...
- Why use `addon.keys.onHotKeyChange(cb(key_id,))` when we already have an event listener for the settings API?
- `addon.keys.onHotKey(cb(key_id))` if this is the only useful one, why not just have it be `addon.self.addEventListener("key", callback)`

**Reply by pufferfish101007** [reply]
Oh right, sorry I didn't realise they were part of the normal setting option thing... that makes sense.

**Reply by pufferfish101007** [reply]
Ok, but I think that tge event listener fits into the `addon.settings` api more than the `addon.self` - `addon.self` is more about the actual addon rather than what's happening in the browser.
